### PR TITLE
Add surface-based backbones

### PR DESCRIPTION
**Pull request by AGrigis**
Roadmap
--------------

- [ ] Add SiT SSL estimator.
- [ ] Release FreeSurger geometry.
- [ ] Release HemiSCNN weigths.
- [x] Add SiT attention maps.
- [x] Add SiT backbone.
- [x] Add SCNN / HemiSCNN backbone.
- [x] Add surfacic operators.
- [x] Add caching context manger.
- [x] Add basic surfacic augmentations.
- [X] Add plotting utilities.
- [x] Add spherical topology handler.

### CODE REVIEW

File: nidl/surface/backbones/scnn.py
Comment (Duplums) [inline nidl/surface/backbones/scnn.py:1]: Regarding spherical CNN, I think the reference paper is Spherical UNet (Zhao et al.) which introduces DiNe convolutions and the CNN backbone for spherical mesh encoding. Most of the papers (including SiT) are compared to this reference. I think it makes sense to provide an implementation of this backbone first before providing a derived backbone (like this one).

File: nidl/surface/backbones/sit.py [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+    --------
+    >>> from nidl.surface.backbones import SiT
+    >>> model = SiT(
+    ...     order=5,
+    ... )
+    >>> print(model)
+
+    References
+    ----------
+    .. [1] Dahan, Simon et al., Surface Vision Transformers: Attention-Based
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:107]: quote the TMI paper: https://arxiv.org/pdf/2204.03408

File: nidl/surface/backbones/sit.py [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+        -------
+        Self
+            Model with loaded weights.
+
+        Raises
+        ------
+        ValueError
+            If the pretrained model in not recognize.
+        """
+        if name == "vit_tiny":
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:280]: I think I'd create dedicated function "sit_tiny", "sity_small".... that instantiates the correct SiT directly as in the original paper (like resnet18, resnet50 or vit_small_patch16_128...).

File: nidl/surface/backbones/sit.py [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+            )
+        else:
+            state = torch.load(
+                weights_file, map_location="cpu", weights_only=False
+            )
+        model.load_state_dict(state)
+        return model
+
+    @classmethod
+    def _load_weights_imagenet(
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:355]: Maybe it'd be better to load the imagenet weights directly without having to specify the state/imagenet_state (the backbone could be inferred from cls I assume and then everything could be fetched elsewhere?). I'm thinking something similar with https://docs.pytorch.org/vision/main/models/generated/torchvision.models.resnet18.html where "weights" is a simple arg parameter to give and everything is handled by torchvision.

File: nidl/surface/backbones/scnn.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+                order=order,
+                standard_ico=standard_ico,
+                pooling_type="mean",
+                n_jobs=n_jobs,
+            ),
+        )
+
+    def forward(
+            self,
+            x: tuple[torch.Tensor, torch.Tensor],
```
Comment (Duplums) [inline nidl/surface/backbones/scnn.py:271]: I think that model's input format consistency across surface backbones is important. For volume backbones, `x` is always assumed to be of shape [B, C, H, W, D], no matter the backbone (the rest is handled internally). I think we need a similar contract for surface data, otherwise each backbone comes with its own flavor.

File: nidl/surface/backbones/sit.py [hunk 5]
```diff
@@ -0,0 +1,779 @@
...
+
+        # Prediction head
+        if n_classes > 0:
+            state["mlp_head.0.weight"] = imagenet_state["norm.weight"].data
+            state["mlp_head.0.bias"] = imagenet_state["norm.bias"].data
+
+        return state
+
+
+class SiTAttentionMaps(nn.Module):
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:445]: I don't understand this class. It is never used in the SiT backbone ?

File: nidl/surface/backbones/sit.py [hunk 6]
```diff
@@ -0,0 +1,779 @@
...
+
+    Parameters
+    ----------
+    model : torch.nn.Module
+        Model from which attention maps will be extracted.
+    names : tuple[str]
+        List of module names or wildcard patterns to hook.
+        Default ('*attend').
+
+    Returns
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:457]: this class does not actually return anything, it feels like it could be more a function than a class ? Also, why the returned type is numpy.ndarray but everything else is torch.Tensor?

File: nidl/surface/backbones/sit.py [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+            attention_texture[indices] = attentions[idx]
+
+        # Apply mask
+        if mask is not None:
+            attention_texture *= mask[:, None, None]
+
+        return attention_texture
+
+
+class FeedForward(nn.Module):
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:560]: I'd take the same naming convention 'Mlp' as in VisionTransformer from timm to be consistent with the ViT3D implementation. Also, I'd try to be as close as possible of their API.

File: nidl/surface/backbones/sit.py [hunk 8]
```diff
@@ -0,0 +1,779 @@
...
+            self,
+            x: torch.Tensor,
+        ) -> torch.Tensor:
+        """
+        Apply the feed-forward transformation.
+        """
+        return self.net(x)
+
+
+class Attention(nn.Module):
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:614]: Same as above, I'd be consistent with timm interface

File: nidl/surface/backbones/sit.py [hunk 9]
```diff
@@ -0,0 +1,779 @@
...
+
+        # Weighted sum of values
+        out = torch.matmul(attn, v)
+        out = torch.swapdims(out, 1, 2)
+        out = torch.flatten(out, start_dim=2)
+
+        return self.to_out(out)
+
+
+class Transformer(nn.Module):
```
Comment (Duplums) [inline nidl/surface/backbones/sit.py:708]: same as above, I'd be consistent with timm interface.

File: nidl/surface/transforms/intensity.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+from ..nn import (
+    IcoConv,
+)
+from .utils import (
+    Interval,
+    validate_data,
+)
+
+
+class RandomGaussianBlur:
```
Comment (Duplums) [inline nidl/surface/transforms/intensity.py:31]: To be consistent with volume transformation, I'd create a specific SurfaceTransform that derives from Transform and I'd set explicitly the contract for the input data format (what is expected). Everything should derive from it. Also, the nidl transform expects either torch.Tensor or numpy.ndarray as input and always returns a consistent output (same dtype, same device)

Comment (Duplums) [inline nidl/surface/transforms/intensity.py:31]: Finally, the current API swapped surface/transforms --> transforms/surface to be more user-friendly.

File: nidl/surface/transforms/utils.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+"""
+
+from functools import wraps
+from typing import Callable
+
+import numpy as np
+import torch
+
+
+class Interval:
```
Comment (Duplums) [inline nidl/surface/transforms/utils.py:20]: I'd say this is not specific to surface transform but could be used at the transform level.

File: nidl/surface/transforms/utils.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+            return np.random.randint(self.low, self.high + 1)
+        elif self.dtype == float:
+            return np.random.uniform(self.low, self.high)
+        else:
+            raise ValueError(
+                f"Data type not supported: {self.dtype}"
+            )
+
+
+def validate_data(allowed_dims: int | list[int]) -> Callable:
```
Comment (Duplums) [inline nidl/surface/transforms/utils.py:91]: this should be in the SurfaceTransform contract

File: nidl/surface/nn/modules.py [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+import collections
+
+import numpy as np
+import torch
+from torch import nn
+
+from .. import Ico
+
+
+class IcoConv(nn.Module):
```
Comment (Duplums) [inline nidl/surface/nn/modules.py:22]: I'd put all these utility classes in backbones/surface/modules.py.

File: nidl/surface/base.py [hunk 14]
```diff
@@ -0,0 +1,1165 @@
...
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any, Callable, ClassVar
+
+import networkx as nx
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+
+
+class Ico:
```
Comment (Duplums) [inline nidl/surface/base.py:25]: I think the documentation should be improved a bit to understand the necessity of this module/class, e.g. 

"""Icosahedron structure.

Cortical surface data has no fixed vertex count and no regular neighbor
structure across subjects, so the convolution / pooling / patching
operations used in Euclidean deep learning (2D CNNs, ViT) don't directly
apply. The standard workaround in surface-based deep learning is to
resample every subject's cortical data onto a common regular grid: a
subdivided icosahedron projected onto a sphere.

An icosahedron of subdivision order k has 10 * 4**k + 2 vertices, almost
all of them with exactly 6 neighbors (the 12 original icosahedron vertices
keep 5). Because this structure is identical for every subject, it lets you
precompute, once, fixed index arrays that stand in for the "kernel windows"
of a 2D CNN -- there is no sliding window on a mesh, so gather-by-index is
what replaces it.

This class does not perform any learning itself. It only builds and caches
the index arrays that model layers use to gather vertex features before
a learnable combination (linear layer, weighted sum, attention) is applied:

    Need in your model                    Method            Used by
    -------------------------------------------------------------------------
    Base grid / vertex coordinates        Ico.mesh()         all backbones
    Group vertices into ViT patches       Ico.patches()      Surface Vision
                                                               Transformer (SiT)
    Per-vertex neighbor window            Ico.neighbors()    Spherical U-Net,
                                                               MoNet, ChebNet/GCN
    Downsample to a coarser icosphere     Ico.downsample()   U-Net encoder /
                                                               pooling
    Upsample back to a finer icosphere    Ico.interpolate()  U-Net decoder /
                                                               unpooling

Example
-------
>>> order = 6                                    # 40,962 vertices
>>> verts, tris = Ico.mesh(order=order)          # grid all subjects are resampled onto

>>> # ViT-style tokens (SiT / MS-SiT)
>>> patch_idx = Ico.patches(order=order, size=4)
>>> tokens = x[patch_idx]                        # x: [n_vertices, C] -> [n_patches, verts_per_patch, C]

>>> # CNN-style convolution window (Spherical U-Net / MoNet)
>>> neigh_idx = Ico.neighbors(order=order, depth=1, direct_neighbor=True)
>>> window = x[neigh_idx]                        # [n_vertices, 7, C] -> weighted sum with a learnable kernel

>>> # U-Net pooling (encoder) / unpooling (decoder)
>>> down_idx = Ico.downsample(source_order=order, target_order=order - 1)
>>> x_pooled = x[down_idx]
>>> up_idx = Ico.interpolate(source_order=order - 1, target_order=order)
>>> x_unpooled = x_pooled[up_idx].mean(axis=1)

All public methods are cached (see `Ico.cached`): the index arrays depend
only on mesh order/parameters, never on subject data, so they are computed
once per configuration and reused for the entire dataset and training run.

**Review by Duplums — CHANGES_REQUESTED** [review]
Very good v0, some points need to be discussed/updated especially to be consistent with NIDL's API on volume data. Also, to me the priority of the backbones/models should be:
1) Spherical UNet
2) SiT
3) SCNN
4) MS-SiT (last of Robinson)
5) ChebNet/GCN (classical approach from Bronstein)
